Source/WebCore/html/ImageDataArray.h

```diff
@@ -28,34 +28,147 @@
+        case JSC::TypeFloat16: return ImageDataStorageFormat::Float16;
+        default: RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+        }
+    }
+
+    template <typename F>
+    static auto visit(JSC::ArrayBufferView& bufferView, F&& f)
+    {
+        // Always try Uint8ClampedArray first, as it should be the most frequent.
+        if (auto* array = dynamicDowncast<JSC::Uint8ClampedArray>(bufferView))
+            return std::forward<F>(f)(*array);
+        if (auto* array = dynamicDowncast<JSC::Float16Array>(bufferView))
+            return std::forward<F>(f)(*array);
+        RELEASE_ASSERT_NOT_REACHED("Unexpected ArrayBufferView type");
+    }
+
+    template <typename F>
+    auto visit(F&& f) const
+    {
+        return visit(m_bufferView.get(), std::forward<F>(f));
+    }
+
+    Ref<JSC::Uint8ClampedArray> asUint8ClampedArray() const;
+    Ref<JSC::Float16Array> asFloat16Array() const;
+
+    JSC::ArrayBufferView& getArrayBufferView() const { return m_bufferView.get(); }
+    JSC::ArrayBufferView* operator->() const { return &getArrayBufferView(); }
+
     auto length() const
     {
-        Ref bufferView = m_bufferView;
-        return bufferView->length();
+        return visit([](const auto& array) { return array.length(); });
     }
 
-    // FIXME: Remove these temporary functions.
-    Uint8ClampedArray& get() const { return m_bufferView.get(); }
-    auto data() const
+    // Needed by `toJS<IDLUnion<IDLUint8ClampedArray, ...>, const ImageDataArray&>()`
+    using Variant = std::variant<RefPtr<JSC::Uint8ClampedArray>, RefPtr<JSC::Float16Array>>;
+    operator Variant() const
     {
-        Ref bufferView = m_bufferView;
-        return bufferView->data();
+        return visit([](auto& a) {
+            return Variant(RefPtr(&a));
+        });
     }
 
 private:
-    // FIXME: Handle other typed arrays.
-    Ref<JSC::Uint8ClampedArray> m_bufferView;
+    ImageDataArray(Ref<JSC::ArrayBufferView>&& arrayBufferView)
+        : m_bufferView(WTFMove(arrayBufferView))
+    {
+        ASSERT(isSupported(m_bufferView.get()));
+    }
+
+    static std::optional<ImageDataArray> tryCreate(size_t length, ImageDataStorageFormat storageFormat, bool zeroFill)
+    {
+        std::optional<ImageDataArray> array;
+        switch (storageFormat) {
+        case ImageDataStorageFormat::Uint8:
+            if (RefPtr bufferView = JSC::Uint8ClampedArray::tryCreateUninitialized(length)) {
+                if (zeroFill)
+                    bufferView->zeroFill();
+                array.emplace(bufferView.releaseNonNull());
+            }
+            break;
+        case ImageDataStorageFormat::Float16:
+            if (RefPtr bufferView = JSC::Float16Array::tryCreateUninitialized(length)) {
+                if (zeroFill)
+                    bufferView->zeroFill();
+                array.emplace(bufferView.releaseNonNull());
+            }
+        }
+        return array;
+    }
```

Comment:
Why do we keep these long functions in the header file? This makes the interface of the class harder to see.